ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -31,6 +45,245 @@ let tdb_tool = !Xapi_globs.tdb_tool
+        parse_string ~consume:All p x
+    end in
+    let ldap fmt = fmt |> Printf.kprintf @@ Printf.sprintf "ldap %s" in
+    let* kvps = P.parse stdout <!> ldap "parsing failed '%s'" in
+    let get_string k =
+      match Map.find_opt k kvps with
+      | None ->
+          Error (ldap "missing key '%s'" k)
+      | Some x ->
+          Ok x
+    in
+    let get_int of_string k =
+      let* str = get_string k in
+      try Ok (of_string str)
+      with _ -> Error (ldap "invalid value for key '%s'" k)
+    in
+    let* upn = get_string "userPrincipalName" in
+    let* user_account_control = get_int Int32.of_string "userAccountControl" in
+    let* account_expires = get_int Int64.of_string "accountExpires" in
+    let account_expired =
+      (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-accountexpires *)
+      let windows_nt_time_to_unix_time x =
+        Int64.sub (Int64.div x 10000000L) 11644473600L
+      in
+      match account_expires with
+      | 0L | 9223372036854775807L ->
+          false
+      | i ->
+          let expire_unix_time =
+            windows_nt_time_to_unix_time i |> Int64.to_float
+          in
+          expire_unix_time < Unix.time ()
+    in
+    Ok
+      {
+        upn
+      ; account_expired
+        (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-useraccountcontrol
+         * for bit flag docs *)
+      ; account_disabled= Int32.logand user_account_control 2l <> 0l
+      ; account_locked= Int32.logand user_account_control 16l <> 0l
+      ; password_expired= Int32.logand user_account_control 8388608l <> 0l
+      }
+
+  let query_user sid =
+    let query = Printf.sprintf "(&(objectClass=user)(objectSid=%s))" sid in
+    let* stdout =
+      call_ldap query <!> fun _ -> generic_ex "ldap query failed: '%s'" query
+    in
+    parse_user stdout <!> generic_ex "%s"
+end
+
+module Wbinfo = struct
+  let can_resolve_krbtgt () =
+    match call_wbinfo ["-n"; krbtgt] with Ok _ -> true | Error _ -> false
+
+  let sid_of_name name =
+    (* example:
+     *
+     * $ wbinfo -n user@domain.net
+       S-1-2-34-... SID_USER (1)
+     * $ wbinfo -n DOMAIN\user
+       # similar output *)
+    let args = ["-n"; name] in
+    let* stdout =
+      call_wbinfo args <!> fun _ ->
+      generic_ex "'wbinfo %s' failed" (String.concat " " args)
+    in
+    match String.split_on_char ' ' stdout with
+    | sid :: _ ->
+        Ok (String.trim sid)
+    | [] ->
+        Error
+          (generic_ex "unable to find SID in output of 'wbinfo %s'"
+             (String.concat " " args))
+
+  let name_of_sid sid =
+    (* example:
+     * $ wbinfo -s S-1-5-21-3143668282-2591278241-912959342-502
+       CONNAPP\krbtgt 1 *)
+    let args = ["-s"; sid] in
+    let* stdout =
+      call_wbinfo args <!> fun _ ->
+      generic_ex "'wbinfo %s' failed" (String.concat " " args)
+    in
+    (* we need to drop everything after the last space *)
+    match List.rev (String.split_on_char ' ' stdout) with
```

Comment:
It might be easier to use regular expressions for problems like this.

Reply:
You're right, this is much cleaner